tfprovider/provider_test.go

```diff
@@ -0,0 +1,174 @@
+package tfprovider
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	providerInstallPath = "../hack/.terraform/plugins/linux_amd64/"
+)
+
+type providerTestDef struct {
+	name         string
+	requiredEnvs map[string]string
+	configBody   string
+	version      string
+}
+
+// Ensure you run ./hack/init.sh to
+// install the providers before running these tests
+var testedProviders = []providerTestDef{
+	{
+		name:         "aws",
+		requiredEnvs: map[string]string{"AWS_REGION": "us-east-1"},
+		version:      "2.70.0",
+	},
+	{
+		name:         "azurerm",
+		requiredEnvs: map[string]string{},
+		configBody:   `features {}`,
+		version:      "2.22.0",
+	},
+	{
+		name:         "helm",
+		requiredEnvs: map[string]string{},
+		version:      "1.2.3",
+	},
+}
+
+func Test_PrepareProviderConfigWithDefaults_expectNoError(t *testing.T) {
+	for _, tt := range testedProviders {
+		t.Run(tt.name, func(t *testing.T) {
+			// Clear the HCL config env if already set by previous test
+			err := os.Setenv("PROVIDER_CONFIG_HCL", "")
+			if err != nil {
+				panic(err)
+			}
+
+			// Set required envs
+			for name, value := range tt.requiredEnvs {
+				err = os.Setenv(name, value)
+				if err != nil {
+					panic(err)
+				}
+			}
+			provider, err := getInstanceOfProvider(tt.name, providerInstallPath)
+			if err != nil {
+				t.Errorf("failed to get instance of provider. error = %v", err)
```

Comment:
what's the best way to handle returned errors here? Seems we have panic, just log and log and return